transformer_engine/common/utils.py

```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+"""The utilities for Transformer Engine"""
+import inspect
+import warnings
+from enum import Enum
+
+warnings.simplefilter('default')
+
+
+class DeprecatedEnum:    # pylint: disable=too-few-public-methods
+    """DeprecatedEnum"""
+
+    def __init__(self, enum_cls, msg):
+        self.enum_cls = enum_cls
+        self.msg = msg
+
+    def __getattr__(self, name):
+        if name in self.enum_cls.__members__:
+            warnings.warn(self.msg, DeprecationWarning)
+            return self.enum_cls.__members__[name]
+        raise AttributeError(f"{self.enum_cls} does not contain {name}")
```

Comment:
This doesn't feel quite right to me. Usually when interacting with Python enums we are accessing class members, but this changes it so we interact with the members of a specific instance of `DeprecatedEnum`. It seems we are kind of reimplementing metaclasses here, and doing it "properly" would require delving into the depths of advanced Python features...

Not a blocker BTW. We don't need to be 100% robust since this is a backward compatibility convenience, and I suspect doing it "right" would take more effort than it's worth.

Reply:
I agree this is kind of a workaround since I have not found any good approaches to set MetaClass after classes defination.
What we need here is
1. Give deprecated warning when using `Enum` via deprecated way (`transformer_engine.jax.TransformerLayerType` in this case).
2. No impact to usage.

Reimplementing metaclasses can break the requirement 1, for example:
```python
class DeprecatedEnumMeta(type):
    def __new__(cls, name, bases, dct):
        warnings.warn(self.msg, DeprecationWarning)
        # Adding TransformerLayerEnum's members into dct

class DeprecateEnum(metaclass=DeprecatedEnumMeta):
    pass
```
The warning message would be showed right after `DeprecateEnum` defined, instead of invoking (`DeprecateEnum.xxxx`), which is equivalent to give warning in importing stage.

To my knowledge, there is no way to catch the access of class's members (something like `__getattr__` but for classes not instances). Then I implent this as a workaround.

For no impact to usage, it would be great to implement `__iter__()`, like
```python
class DeprecatedEnum:    # pylint: disable=too-few-public-methods
    """DeprecatedEnum"""

    def __init__(self, enum_cls, msg):
        self.enum_cls = enum_cls
        self.msg = msg

    def __iter__(self):
        return iter(list(self.enum_cls.__members__.values()))

    def __getattr__(self, name):
        if name in self.enum_cls.__members__:
            warnings.warn(self.msg, DeprecationWarning)
            return self.enum_cls.__members__[name]
        raise AttributeError(f"{self.enum_cls} does not contain {name}")

>>> import transformer_engine.jax as te
>>> te.TransformerLayerType.DECODER in te.TransformerLayerType
transformer_engine/common/utils.py:22: DeprecationWarning: TransformerLayerType is moving to transformer_engine.jax.flax module
  warnings.warn(self.msg, DeprecationWarning)
True
>>> te.TransformerLayerType.DECODER is te.flax.TransformerLayerType.DECODER
True
```

Reply:
Does this make sense?

Reply:
The handling for `DeprecatedEnum` is starting to get messy. I'm not very familiar with this level of esoteric Python, but I wonder if the following would work:
```python
def deprecate_wrapper(obj, msg):
    """Deprecate wrapper"""

    def _warn():
        warnings.warn(msg, DeprecationWarning)

    if inspect.isclass(obj):

        class DeprecatedMetaclass(type):
            """Metaclass for deprecated class wrapper"""
            def __getattribute__(cls, name):
                _warn()
                return super().__getattribute__(cls, name)

        class DeprecatedClass(obj, metaclass=DeprecatedMetaclass):
            """Deprecated class wrapper"""
            def __init__(self, *args, **kwargs):
                _warn()
                super().__init__(*args, **kwargs)

        return DeprecatedClass

    if inspect.isfunction(obj):

        def deprecated(*args, **kwargs):
            _warn()
            return obj(*args, **kwargs)

        return deprecated

    raise NotImplementedError(
        f"deprecate_wrapper only support Class and Function, but got {type(obj)}.")

```
Thoughts:
- No need to treat enums different than other classes.
- Minimal function overriding and no extra state.
- Printing the warning in a helper function prevents redundant messages if we both create a class instance and access a class variable.
- I am not sure how this will handle the case where you have multiple deprecated classes. I think `warnings` will filter based on line number, so it will probably only complain about the first deprecated class you use.

This is mostly for personal interest and isn't a serious suggestion.

Reply:
The potential issue of above suggestion is that Enum in Python is not allowed to extend. If we define a class to inherit a Enum, then we get `cannot extend enumeration xxxx`. Let's discuss offline to figure out a better approach for improvment.

Reply:
Interesting, that does complicate things. I guess dealing with enums is going to be messy no matter what then.

Reply:
Note, if it is too complicated for enum, we could just not give deprecation warning in the code and only in the doc.
If they use the old enum, they will also use the old fct and classes and those already give a warning. So they will get warning in all cases.
I wouldn't speed too much time on that now.